include/sleipnir/control/OCPSolver.hpp

```diff
@@ -0,0 +1,205 @@
+using FixedStepConstraintFunction =
+    std::function<void(const VariableMatrix&, const VariableMatrix&)>;
+
+/**
+ * Performs 4th order Runge-Kutta integration of dx/dt = f(t, x, u) for dt.
+ *
+ * @param f  The function to integrate. It must take two arguments x and u.
+ * @param x  The initial value of x.
+ * @param u  The value u held constant over the integration period.
+ * @param t0 The initial time.
+ * @param dt The time over which to integrate.
+ */
+template <typename F, typename State, typename Input, typename Time>
+State RK4(F&& f, State x, Input u, Time t0, Time dt) {
+  auto halfdt = dt * 0.5;
+  State k1 = f(t0, x, u);
+  State k2 = f(t0 + halfdt, x + halfdt * k1, u);
+  State k3 = f(t0 + halfdt, x + halfdt * k2, u);
+  State k4 = f(t0 + dt, x + dt * k3, u);
+
+  return x + (k1 + k2 * 2.0 + k3 * 2.0 + k4) * (dt / 6.0);
+}
+
+/**
+ * Enum describing an OCP transcription method.
+ */
+enum TranscriptionMethod {
+  kDirectTranscription,
+  kDirectCollocation,
+  kSingleShooting
+};
+
+/**
+ * Enum describing a type of system dynamics constraints.
+ */
+enum DynamicsType { kExplicitODE, kDiscrete };
+
+/**
+ * This class allows the user to pose and solve a constrained optimal control
+ * problem (OCP) in a variety of ways.
+ *
+ * The system is transcripted by one of three methods (direct collocation,
+ * direct transcription, or single-shooting) and additional constraints can be
+ * added.
+ *
+ * @todo expand this explanation more.
+ * In single-shooting, states depend explicitly as a function of all previous
+ * states and all previous inputs. In direct transcription, each state is a
+ * decision variable constrained to the integrated dynamics of the previous
+ * state. In direct collocation, the trajectory is modeled as a series of cubic
+ * polynomials where the centerpoint slope is constrained.
+ *
+ * Explicit ODEs are integrated using RK4.
+ *
+ * For explicit ODEs, the function must be in the form dx/dt = f(t, x, u).
+ * For discrete state transition functions, the function must be in the form
+ * xₖ₊₁ = f(t, xₖ, u).
+ *
+ * Direct collocation requires an explicit ODE. Direct transcription and
+ * single-shooting can use either an ODE or state transition function.
+ */
+class SLEIPNIR_DLLEXPORT FixedStepOCPSolver : public OptimizationProblem {
+ public:
+  /**
+   * Build an optimization problem using a system evolution function (explicit
+   * ODE or discrete state transition function).
+   *
+   * @param numStates The number of system states.
+   * @param numInputs The number of system inputs.
+   * @param dt The timestep for integration. Must be in seconds.
+   * @param numSteps The number of control points.
+   * @param dynamics The system evolution function, either an explicit ODE or a
+   * discrete state transition function.
+   * @param dynamicsType The type of system evolution function.
+   * @param method The transcription method.
+   */
+  FixedStepOCPSolver(int numStates, int numInputs, double dt, int numSteps,
```

Comment:
Use std::chrono::seconds for times. We could also consider importing the units library from the tests if that's more ergonomic.

Reply:
`duration<double>` seems like the right thing here, since seconds seems to work only as an integer count.